### DISCUSSION TITLE
Potential N+1 query issue with capture_exception in Sentry SDK for Django after upgrading to 2.17.0

### DISCUSSION THREAD

**Discussion by gabicavalcante** [getsentry/sentry-python#3718] [Q&A]
Hi everyone! I recently upgraded from sentry-sdk version 1.4.3 to 2.17.0 for my Django project and noticed a curious behavior. When I use capture_exception with certain exceptions, it seems to kick off extra database queries, especially when Django model instances are involved. It’s almost like it’s adding a bit of an N+1 query pattern, which I hadn’t seen in the older version.

Here’s a little example of what I’m seeing:

```python 
from sentry_sdk import capture_exception

def process_document(instance_a, keyname):
    try:
        document = instance_a.documents.get(keyname=keyname)
        # more processing
    except DocumentInfo.DoesNotExist:
        LOGGER.error("No DocumentInfo found.")
        return None
    except DocumentInfo.MultipleObjectsReturned as e:
        LOGGER.exception("Multiple DocumentInfo entries found.")
        capture_exception(e)  # This seems to trigger extra queries
        return None
```

This is a legacy codebase we’re slowly refactoring, but I found it interesting because this issue wasn’t happening before the upgrade. We actually have an assert_num_queries test to keep an eye on performance, and thanks to that test, we caught the increase: the function jumped from 10 queries to 23 after the upgrade! Our current fix was to keep only the `logger.exception` with the data we think is critical to know about the issue.

Has anyone else noticed this, or maybe found a workaround? It would be great to hear any thoughts! Thanks!

**Comment by sl0thentr0py** [comment]
Strange, I would need more info to help you.

>  and thanks to that test, we caught the increase: the function jumped from 10 queries to 23 after the upgrade

Can you dig a bit more into what kind of query it is that increased? And also make sure that that does not happen if you disable Sentry?

**Reply by gabicavalcante** [reply]
Sure, I tested to remove this call to see which query we were doing; they were like 

```
SELECT "documentinfo"."id" FROM "documentinfo" WHERE "documentinfo"."id" = 'xxxxx'::uuid LIMIT 21
```
Those queries are related to `document = instance_a.documents.get(keyname=keyname)`, it's the reason for the `DocumentInfo.MultipleObjectsReturned`.

With the previous version, they did not happen. I thought maybe because of this upgrade, when capture_exception(e) was invoked, it implicitly serialized the exception data, including model instances, which can lead to unintended database queries as it tries to retrieve values.

**Reply by sl0thentr0py** [reply]
does calling `repr(document)` invoke the query? we moved the serialization of stack variables earlier, but it still did happen in the pipeline before so behaviorally it should stay the same amount of queries. If somehow behavior has changed, that's a regression on our end.

**Reply by sl0thentr0py** [reply]
could you also give me `type(document)` and `type(document).__mro__`?

**Reply by gabicavalcante** [reply]
In my case, this query `document = instance_a.documents.get(keyname=keyname)` will raise an error, so I don't have a document instance when the issue happens because the `get` returns multiple objects. But in case it works, `type(document)` will be something like:

```python
ipdb> type(document)
<class 'django_app.models.DocumentInfo'>
ipdb> type(document).__mro__
(<class 'django_app.models.DocumentInfo'>, <class 'common.models.APINode'>, <class 'common.models.Timestamped'>, <class 'django.db.models.base.Model'>, <class 'django.db.models.utils.AltersData'>, <class 'object'>)
```

Also, adding `repr(document)` doesn't increase the number of queries. 

When the exception happens, I tried to debug and find where the query was made... I gather some outputs: 

```python
---> 75         capture_exception(e)
     76         return None

ipdb> e
MultipleObjectsReturned('get() returned more than one DocumentInfo -- it returned 2!')

> /usr/local/lib/python3.9/site-packages/sentry_sdk/scope.py(1206)capture_exception()
   1205         if error is not None:
-> 1206             exc_info = exc_info_from_error(error)

> /usr/local/lib/python3.9/site-packages/sentry_sdk/utils.py(1075)exc_info_from_error()
   1074     elif isinstance(error, BaseException):
-> 1075         tb = getattr(error, "__traceback__", None)
   1076         if tb is not None:

> /usr/local/lib/python3.9/site-packages/sentry_sdk/scope.py(1211)capture_exception()
   1210         event, hint = event_from_exception(
-> 1211             exc_info, client_options=self.get_client().options
   1212         ) <---- Here, the variable event has the queryset 
```

When I checked the event variable, I have something like this:
```python
{
    'level': 'error',
    'exception': {
        'values': [
            {
                'mechanism': {
                    'type': 'generic',
                    'handled': True
                },
                'module': 'uspto_doc_fetcher.models',
                'type': 'DocumentInfo.MultipleObjectsReturned',
                'value': 'get() returned more than one DocumentInfo -- it returned 2!',
                'stacktrace': {
                    'frames': [
                        {
                            'filename': '...py',
                            'abs_path': '...py',
                            'function': 'process_document',
                            'module': 'module-name',
                            'lineno': 36,
                            'pre_context': [
                                '',
                                'def process_document(instance_a, keyname)',
                                '    try:'
                            ],
                            'context_line': '        document = instance_a.documents.get(keyname=keyname)',
                            'vars': {
                                'e': "MultipleObjectsReturned('get() returned more than one DocumentInfo -- it returned 2!')",
                                'created': 'True',
                                'document_infos': '<QuerySet [<DocumentInfo: pk123>, <DocumentInfo: pk123>]>',
                                'ipdb': "<module 'ipdb' from '/usr/local/lib/python3.9/site-packages/ipdb/__init__.py'>"
                            }
                        },
                        {
                            'filename': 'django/db/models/query.py',
                            'abs_path': '/usr/local/lib/python3.9/site-packages/django/db/models/query.py',
                            'function': 'get',
                            'module': 'django.db.models.query',
                            'lineno': 640, 
                            ...
                            'context_line': '        raise self.model.MultipleObjectsReturned(',
                            'post_context': [
                                '            "get() returned more than one %s -- it returned %s!"',
                                '            % (',
                                '                self.model._meta.object_name,',
                                '                num if not limit or num < limit else "more than %s" % (limit - 1),',
                                '            )'
                            ],
                            'vars': {
                                'self': '<QuerySet [<DocumentInfo: pk123>, <DocumentInfo: pk123>]>',
                                'args': [],
                                'kwargs': {...},
                                'clone': '<QuerySet [<DocumentInfo: pk123>, <DocumentInfo: pk123>]>',
                                'limit': '21',
                                'num': '2'
                            }
                        }
                    ]
                }
            }
        ]
    }
}
```

**Reply by sl0thentr0py** [reply]
hmm, the `QuerySet`s should look like this, I will try to reproduce, thanks for the added info!
https://github.com/getsentry/sentry-python/blob/02d09346e6d070e03b828807d72485b6f23b2c11/sentry_sdk/integrations/django/__init__.py#L263-L267

**Reply by sl0thentr0py** [reply]
we actually have a test covering this branch, so I'd need more details from you for reproduction
https://github.com/getsentry/sentry-python/blob/02d09346e6d070e03b828807d72485b6f23b2c11/tests/integrations/django/test_basic.py#L292-L310

* python version
* django version
* anything else that's potentially different about your setup, especially regarding database models and if you have custom behaviour implemented around `QuerySet`s

**Reply by gabicavalcante** [reply]
* python version: 3.9
* django version: 4.2.16
* anything else that's potentially different about your setup, especially regarding database models and if you have custom behavior implemented around QuerySets: _I don't think we have any other diff that could impact on it; it's a really simple model (models.Model), without any custom queryset behavior_

I'm going to debug a little more to see exactly where this queryset is evaluated.

**Reply by gabicavalcante** [reply]
I still couldn't find the root cause of those queries :( I made a simpler test and got 12 queries 👀 

```python
        with django_assert_num_queries(1):
            try:
                document =instance_a.documents.get()
            except Exception as e:
                from sentry_sdk import capture_exception
                capture_exception(e)
```

It's not the `_django_queryset_repr` because when I tried to use only this method in the test, no more queries were executed.

**Reply by sl0thentr0py** [reply]
the problem is that it should use `_django_queryset_repr` but it's instead using the normal `repr` in your case.

**Reply by sl0thentr0py** [reply]
can you give me
* the whole model class, or a simpler copy
*  `django_assert_num_queries`

so that I can try to reproduce
